### PR TITLE
Add ctx to `beforeSSR` and return 404 error for 404 pages

### LINKED ISSUE #264: Return 404 status for 404 pages
**_Bug report_**

- [x] I've searched for existing issues
- [x] I've checked documentation: https://docs.frontity.org
- [x] I've checked community forum: https://community.frontity.org

**Describe the expected behavior**

404 pages should return a 404 status code.

**Describe the observed behavior**

Right now they return a 200 status code.

**Describe the steps involved to reproduce the problem**

- Visit a page that doesn't exist: https://mars.frontity.org/doesnt-exist.
- It returns a 200 status code.

**Possible solution**

This is tricky because we need to change the response object from the React render. Maybe we can pass Koa's `ctx` to the `init` and `beforeSSR` methods and `tiny-router` can check if the data of the current `link` is `is404` and change status to 404 using `ctx.status = 404`:

```js
export const beforeSSR: TinyRouter["actions"]["router"]["beforeSSR"] = async ({
  state,
  actions
}) => ctx => {
  if (state.router.autoFetch) {
    if (actions.source && actions.source.fetch) {
      await actions.source.fetch(state.router.link);
      const data = state.source.get(state.router.link);
      if (data.isError)
         ctx.status = data.errorStatus;
    } else
      console.warn(
        "You are trying to use autoFetch but no source package is installed."
      );
  }
};
```

**Dependencies**

- https://community.frontity.org/t/expose-koas-context-in-the-beforessr-afterssr-methods/1067.
- https://community.frontity.org/t/add-status-code-and-error-messages-to-data/1125

### LINKED ISSUE #306: Add Koa's ctx to beforeSSR argument
**Feature:** [Expose Koa's context in the beforeSSR method](https://community.frontity.org/t/expose-koas-context-in-the-beforessr-method/1067)
**User Story**: Change Koa's response object in my beforeSSR method

- Add `ctx` to `beforeSSR`.
- Update the docs (we can add it here https://docs.frontity.org/learning-frontity/actions#actions-triggered-by-frontity)

[Sprint Card](https://github.com/orgs/frontity/projects/2#card-32781566)

### PR DESCRIPTION
**Pull request by michalczaplinski**
#### Description of what you did:
Add `ctx` to `beforeSSR` and return correct HTTP  status code in frontity for corresponding errors from a server.

- [x] Add `ctx` to `beforeSSR`.
- [x] Update the docs (we can add it here https://docs.frontity.org/learning-frontity/actions#actions-triggered-by-frontity)

fixes: #306 
fixes: #264 

[Documentation](https://app.gitbook.com/@frontity-tech/s/frontity/~/diff/drafts/-M00Ls29V3QLo_LSrcue/@drafts)

#### My PR is a:
- 🔝 Improvement

#### Is the PR ready to be merged?
- ✅ Done

### CODE REVIEW

File: packages/core/src/server/index.tsx [hunk 1]
```diff
@@ -79,7 +79,7 @@ export default ({ packages }): ReturnType<Koa["callback"]> => {
     // Run beforeSSR actions.
     await Promise.all(
       Object.values(store.actions).map(({ beforeSSR }) => {
-        if (beforeSSR) return beforeSSR();
+        if (beforeSSR) return beforeSSR(ctx);
```
Comment (luisherranz) [inline packages/core/src/server/index.tsx]: I would make this an object, just in case we need to add more stuff in the future:
```js
beforeSSR({ ctx });
```

File: packages/tiny-router/src/actions.ts [hunk 2]
```diff
@@ -43,16 +44,24 @@ export const init: TinyRouter["actions"]["router"]["init"] = ({
   }
 };
 
-export const beforeSSR: TinyRouter["actions"]["router"]["beforeSSR"] = async ({
+export const beforeSSR: TinyRouter["actions"]["router"]["beforeSSR"] = ({
   state,
   actions
-}) => {
+}) => async (
+  ctx: Koa.ParameterizedContext<Koa.DefaultState, Koa.DefaultContext>
```
Comment (luisherranz) [inline packages/tiny-router/src/actions.ts]: For TypeScript I think I would expose a `ServerAction` in `"@frontity/types"` with our own `Context` type because it will include Koa's context but also more things in the future (like `ctx.settings`, `ctx.template` and so on...).

```ts
import { ServerAction } from "frontity/types";

actions: {
  myPackage: {
    beforeSSR: ServerAction<MyPackage>;
  ...
}
```

That way, people can launch other server actions from the `beforeSSR` one and organize their code better:

```ts
actions: {
  myPackage: {
    beforeSSR: ServerAction<MyPackage>;
    doSomethingWithContext: ServerAction<MyPackage>;
    doSomethingElseWithContext: ServerAction<MyPackage>;
  ...
}
```
```js
actions: {
  myPackage: {
    beforeSSR: ({ actions }) => async ({ ctx }) => {
      await Promise.all([
        actions.myPackage.doSomethingWithContext({ ctx }),
        actions.myPackage.doSomethingElseWithContext({ ctx })
      ]);
    }
  }
}
```

File: packages/tiny-router/src/index.ts [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 
 const tinyRouter: TinyRouter = {
   name: "@frontity/tiny-router",
   state: {
     router: {
       link: "/",
       autoFetch: true
-    }
+    },
+    source: WPSource().state.source // TODO: notr sure if that's correct
```
Comment (luisherranz) [inline packages/tiny-router/src/index.ts]: It's not correct. The state from other packages is available on runtime.

**Review by luisherranz — COMMENTED** [review]
I know this is not ready for review yet, but I wanted to give you my impressions :)

File: packages/tiny-router/src/actions.ts [hunk 4]
```diff
@@ -43,16 +44,24 @@ export const init: TinyRouter["actions"]["router"]["init"] = ({
...
   if (state.router.autoFetch) {
-    if (actions.source && actions.source.fetch)
+    if (actions.source && actions.source.fetch) {
       await actions.source.fetch(state.router.link);
-    else
+    }
+
+    const data = state.source.get(state.router.link);
+    if (data.isError) {
+      ctx.status = data.errorStatus;
```
Comment (luisherranz) [inline packages/tiny-router/src/actions.ts]: Hey @michalczaplinski, I guess you're solving this issue in this PR as well, right? https://github.com/frontity/frontity/issues/264

If so, could you please link the issue, assign it to you and move it to in progress? Thanks!

Comment (michalczaplinski) [reply packages/tiny-router/src/actions.ts]: yup, done.

File: packages/tiny-router/src/index.ts (hunk 3 above)
Comment (michalczaplinski) [reply packages/tiny-router/src/index.ts]: gotcha 👍

**Comment by luisherranz** [comment]
It looks like the problem in the e2e test was that Cypress throws for `cy.request` and `cy.visit` unless you add a `failOnStatusCode: false` option. Check:
- https://docs.cypress.io/api/commands/visit.html#Arguments
- https://docs.cypress.io/api/commands/request.html#Arguments

----

Apart from that, I was not able to reproduce the problem you describe in [this video](https://www.loom.com/share/6eb65393bd6d43259c365bab2ff2489c). For me, `await actions.source.fetch(...)` works fine. Take a look at the e2e test: https://github.com/frontity/frontity/blob/917d34bdafc3bca9a969ca4165e70205de42ef98/e2e/packages/wp-source-errors/src/index.tsx#L32-L34

**Comment by luisherranz** [comment]
I have fixed the types of `tiny-router` which were missing `state.source.get` and I have merged `dev` because this branch was out of sync.

File: e2e/project/frontity.settings.ts [hunk 5]
```diff
@@ -35,11 +35,7 @@ const settings: Settings = [
   },
   {
     name: "wp-source-errors",
-    packages: [
-      "e2e-wp-source-errors",
-      "@frontity/wp-source",
-      "@frontity/tiny-router"
-    ]
+    packages: ["e2e-wp-source-errors", "@frontity/wp-source"]
```
Comment (michalczaplinski) [inline e2e/project/frontity.settings.ts]: @luisherranz I'm a bit confused. Why did you remove the `tiny-router`? 

I've added it because I needed to test that the `beforeSSR` *in the* tiny-router package adds the correct error message.

File: e2e/packages/wp-source-errors/src/index.tsx [hunk 6]
```diff
@@ -28,8 +28,10 @@ const wpSourceErrors: WpSourceErrors = {
           state.source.api = "https://test.frontity.io/wp-json";
         }
       },
-      beforeSSR: async ({ state, actions }) => {
+      beforeSSR: ({ state, actions }) => async ctx => {
```
Comment (michalczaplinski) [inline e2e/packages/wp-source-errors/src/index.tsx]: @luisherranz This was not relevant for my tests, because I needed to rely on the `beforeSSR` of the `tiny-router`. 

The strange thing is that with your changes, I can comment out the content of the `beforeSSR` in this file and the tests pass **and** I get the correct errors when I navigate to the page manually with my browser.

**Review by michalczaplinski — COMMENTED** [review]
@luisherranz I don't fully understand why this is working now and was not working before. 

What I did on top of your changes: 
1. I re-added the `tiny-router` to the `wp-source-errors` [here](https://github.com/frontity/frontity/pull/309/commits/917d34bdafc3bca9a969ca4165e70205de42ef98#diff-d35f28fb9a23785741a203fa518e1450R38)
2. I commented out the `beforeSSR` method that you added [here](https://github.com/frontity/frontity/pull/309/commits/917d34bdafc3bca9a969ca4165e70205de42ef98#diff-14adba8e1fc2b93ce0a92fb4a9625e9eR31)

The test pass now (I know that I was missing the `failOnStatusCode: false` as you pointed out, but regardless of that koa was returning 200 when it should have returned an error)

File: e2e/project/frontity.settings.ts (hunk 5 above)
Comment (luisherranz) [reply e2e/project/frontity.settings.ts]: I thought that it was not needed to test that we can change the `ctx.status` from a `beforeSSR` method... but it's fine. Add it back if you want.

File: e2e/packages/wp-source-errors/src/index.tsx (hunk 6 above)
Comment (luisherranz) [reply e2e/packages/wp-source-errors/src/index.tsx]: Oh, do you mean that the tests pass without any code? Maybe the tests are wrong... I don't know. Could you please check them out?

**Comment by luisherranz** [comment]
I am a bit confused about your messages... Is this working? Are the tests correct? Do you need more help?

File: e2e/packages/wp-source-errors/src/index.tsx (hunk 6 above)
Comment (michalczaplinski) [reply e2e/packages/wp-source-errors/src/index.tsx]: I meant that they pass because of this: https://github.com/frontity/frontity/pull/309/files#diff-d8058b9e7fa71c069f1e5c2f6a255444L50-R65

We set the correct status code in the `beforeSSR` action of the `tiny-router`.

**Comment by michalczaplinski** [comment]
Nah, I'm fine now. I still don't know why I was having the issue that I described, but since I cant reproduce it now, I'm just gonna move on.

**Comment by luisherranz** [comment]
Great! 👍

File: packages/types/src/action.ts [hunk 7]
```diff
@@ -42,3 +74,11 @@ export type AsyncAction<Pkg extends Package, Input = null> = [Input] extends [
       actions: ResolveActions<Pkg["actions"]>;
       libraries: Pkg["libraries"];
     }) => (args: Input) => Promise<void>;
+
+type Context = Koa.ParameterizedContext<Koa.DefaultState, Koa.DefaultContext>;
+
```
Comment (michalczaplinski) [inline packages/types/src/action.ts:84]: @luisherranz I think this is correct? We have to support every use case like the following, right?

```js
// SYNC
// 1. sync, no params
action: ({state, ...rest }) => {}

// 2. sync, with params
action: ({state, ...rest }) => params =>  {}

// ASYNC
// 3. async, no params
action: async ({state, ...rest }) =>  {}

// 4. async, with params
action: ({state, ...rest }) => async (params) => {}
```

File: packages/tiny-router/types.ts [hunk 8]
```diff
@@ -8,6 +8,9 @@ interface TinyRouter extends Router {
     router: {
       autoFetch?: boolean;
     };
+    source?: {
+      get?: Source["state"]["source"]["get"];
+    };
```
Comment (michalczaplinski) [inline packages/tiny-router/types.ts:13]: Added cause we use `state.source.get()` inside the `beforeSSR()` in tiny-router

File: packages/types/src/action.ts (hunk 7 above)
Comment (michalczaplinski) [reply packages/types/src/action.ts:84]: oh, I realized that `beforeSSR` is supposed to always be async (at least according to the docs)

**Review by luisherranz — APPROVED** [review]
This is great. Awesome work @michalczaplinski!

By the way, I've added some missing unit tests and removed some things that were not needed. Check my commits and their messages for more information :)

**Comment by luisherranz** [comment]
Docs are great too so I'm going to merge this :)

File: packages/types/src/action.ts (hunk 7 above)
Comment (michalczaplinski) [reply packages/types/src/action.ts:84]: nevermind, that's not true. It can be both.

File: packages/types/src/action.ts [hunk 9]
```diff
@@ -75,12 +77,9 @@ export type AsyncAction<Pkg extends Package, Input = null> = [Input] extends [
...
+  Koa.DefaultState,
+  Koa.DefaultContext
+>;
 
-export type ServerAction<Pkg> =
-  | Action<Pkg>
-  | Action<Pkg, Context>
-  | AsyncAction<Pkg>
-  | AsyncAction<Pkg, Context>;
+export type ServerAction<Pkg> = AsyncAction<Pkg, { ctx: Context }>;
```
Comment (michalczaplinski) [inline packages/types/src/action.ts:85]: @luisherranz You wrote in the commit message: 
> I think that we can leave ServerAction as both async and requiring a Context for the moment.

I thought that it all 4 should be possible:
- sync, no `ctx`
- sync, with `ctx`
- async, no `ctx`
- async, with `ctx`
But with the type that you added, TS will only allow "async with ctx". Is there something I'm missing?

See the comment above for more details.

Comment (luisherranz) [reply packages/types/src/action.ts:85]: Yes, let's leave it like that for the moment. If people want to use other combinations in their packages they can use `Action` and `AsyncAction` when defining the types of their `beforeSSR` action.
